### ISSUE TITLE
App keeps forcing deleting and re-adding of all accounts

### ISSUE THREAD

**Issue by superfrustrated** [robotswingset/narwhal-bugs#22]
**Describe the bug**

- App keeps forgetting my logins, logging me out, however it works. I have five accounts, but mostly only use two of those. On two devices I’m being logged out of all five accounts.  I know this because when I try to upvote I see errors like “Sorry. There was an error saving your vote [or “post” rather than “vote”].  Please try again.”.
- If I yet try to comment: “Please fix the following: - please log in to do that”. 
- If I try to switch to another account, I get “Error. Unable to sign into [account name goes here]. Please try again. If the issue persists, please delete the account and re-authenticate.”
- This already happened twice in the past week or so, and now today a third time. The first time I could (had to) delete all accounts on both devices and log back into all of them, a huge pain. The second time, I deleted all accounts, but each time I tried to re-add in the app crashed during the re-add process, forcing me to delete the app, reinstall, and re-add the five accounts. Same was required on both devices. Extra huge pain. Third time was last night. I have yet to go through the aggravation of trying to delete/re-add on the devices out see if app deletion will be required again or not. But this third time I was also being shown a mixture of subscribed subreddits and random unwanted ones I am not subscribed to, which is more like what I guess I’d expect when being treated by Reddit as a user not logged in, but these weren’t top 10/default subreddits, they were random ones and most I had never heard of before or every visited on any account. 

**To Reproduce**
- The initial failure to recognize that I am already logged into any account has happened spontaneously all three times. I’m a light user, I don’t switch between accounts or comment much, and have no trigger to offer. I don’t know why is happening or what triggers it. 
- Steps to reproduce the behavior after above mass-auto-logout has occurred:
1. Delete all accounts 
2. Try to re-add accounts. Either this works, or doesn’t and forces deleting and reinstalling the app, then re-adding all accounts works. 
3. A few days later, it recurs without warning or reason. In the interim, I can use Reddit normally including seeing only subscribed subreddits, commenting, upvoting, switching between accounts in the app. 

**Expected behavior**
- To log into accounts, stay logged in, be able to use Reddit, and not be regularly/repeatedly forced to delete and re-add accounts, up to and including deleting and reinstalling the app to successfully do so. 

**Screenshots**

![IMG_6351](https://github.com/robotswingset/narwhal-bugs/assets/151955564/7341b27c-c47a-43db-b4e4-2c35a058b915)
![IMG_6349](https://github.com/robotswingset/narwhal-bugs/assets/151955564/26088761-d7d1-4f67-8791-8ee44577b357)
![IMG_6350](https://github.com/robotswingset/narwhal-bugs/assets/151955564/1ac63946-db8a-4a4d-9a97-d08a631206ec)
![IMG_6353](https://github.com/robotswingset/narwhal-bugs/assets/151955564/4463982d-bfd5-4885-b6db-56d23ef82fd1)
![IMG_6352](https://github.com/robotswingset/narwhal-bugs/assets/151955564/cbf7a1c6-a7fa-487b-8b53-3f25b8d51e27)
![IMG_6354](https://github.com/robotswingset/narwhal-bugs/assets/151955564/1a4d512b-4b25-4ef0-abce-780806e16289)

**Device Information:**
- I use the app on two devices, an iPhone and iPad. The iPhone is at iOS 16.7.2 (latest iOS 16) and the iPad is at the latest iOS 17 as of 11/24/23 (not near at hand this moment to get exact version). This happens on both at the same time. When the problem occurs on one, the other has the same problem too when I check. 
- Narwhal is fully updated on both to v2.105 build 929. 

**Additional context**
I’m not a problem user, I don’t draw the attention of mods or Reddit, I’ve never been banned or warned, so I’m confident this has nothing to do with Reddit being upset with me. They’d have to simultaneously be upset with me across five accounts all using different email addresses at once anyway (though traffic on the accounts would be coming from the same IPs, I assume, letting them know they’re all me).

**Comment by ilyanep** [comment]
This is the same issue as #21, though more thoroughly described here.

**Comment by superfrustrated** [comment]
> This is the same issue as #21, though more thoroughly described here.

Thanks, but not quite. My report is a superset of #21, because while I do describe the issue with the app crashing when trying to add accounts like #21, my primary issue is that **the app keeps getting into a state (for reasons unknown) that requires the removal and re-addition of all accounts in the first place**, this having happened three times now in the last week or so on both my devices around the same time.

First time, I was required to delete and re-add all accounts. The second time, I had the #21 issue that required deleting and reinstalling the app before I could re-add the accounts.  Third time I’m just fed up, reading Reddit without voting etc. without wanting to do that deletion/re-adding work again at this moment, even if only account removal and re-addition (without app removal and reinstalling first) will be needed.

**Comment by ilyanep** [comment]
Sorry, I missed the fact that you were able to re-add accounts successfully without the crash bug once. In my case, both things tend to co-occur. I suspect they’re at least related. Hope a fix is able to be made soon!

**Comment by rickharrison** [comment]
Thanks for these bug reports y'all. This was an issue with the Reddit API, but I've been able to engineer around it. Fix coming very soon.

@superfrustrated Are you in testflight? Can you help me verify this fix

**Comment by superfrustrated** [comment]
> Sorry, I missed the fact that you were able to re-add accounts successfully without the crash bug once.

Hey no worries!

**Comment by superfrustrated** [comment]
> @superfrustrated Are you in testflight? Can you help me verify this fix

Thank for your efforts. 

@rickharrison I was in TF, and would like to help, but long story short, I need to ask for a TF reinvite, if you want to tell me how to request that again?

I have more to report that I’ll add in a new comment below though, which you may or may not find helpful. But feel free to tell me where to how to request TestFlight invite again if you like.

**Comment by superfrustrated** [comment]
Trying to be descriptive but concise… not my strong suit.  Sorry. 

So I opened this report and provided screen grabs of errors etc, but had not done the account removal-re-add yet. 

Last night, maybe 12hrs later, I picked up my devices to just read (accepting that would be without voting, commenting, or account switching, which had been generating errors) and found the app working normally and allowing those things without errors, without my having done the account removal/re-add). Did a bunch of commenting, voting, flagging all above as read/hiding all read, basically the app worked normally, no errors. 

12 hours or so later, now, the app’s back throwing errors again, but not consistently. It occurred to me that exiting/rerunning the app might be the reason or help or reset my sick state, so I’ve tried exiting and rerunning a bunch of times and testing voting/commenting. 

I’ve just been opening the app, trying to upvote/comment, and seeing varying results. 

1) Sometimes, I get one vote without error, then a second vote attempt generates the error.  Exiting/re-entering the app generated this result a few times.. First vote or “action” appears to work, second wouldn’t. 

2) On a couple of tries, I made a few votes successfully, then errors returned again (more than one successful “action” before return to sick state in these attempts). 

3) On another exit/re-run, I was able to vote several times, added a comment and then deleted it, saved and then unsaved a post, basically take many actions, and so far no errors are being generated.  Have exited the app, gotten this normal behavior, and exited a few more times and got normal behavior on those subsequent runs of the app. Did this a half-dozen times trying to see if I’d return to the “One action then errors” or “a few actions then errors” states, but note can’t seem to make that happen. HOWEVER, if I try to switch accounts I get the error screenshot above, then the formerly-working-fine session is now in the “errors for any action” state. If I hadn’t tried to switch accounts, that initial session may have stayed functional, but the account switch aren’ attempt broke things again. 

Summary: Without having removed and re-added accounts (which seems to have been the required fix in the past for my screen-shotted errors above), if I exit and re-run the app, sometimes I get one successful “action” before errors resume, sometimes a few actions before errors resume, and now have gotten several sessions in a row on exit/re-run and made a bunch of different actions without errors, but can’t say if more “actions” would cause errors to resume at some point, or if a have a stable/non-error-generating “session”/state here that’ll be fine until I close the app, but trying to switch accounts fails and puts my in the sick state in the “formerly a-ok” initial session until I exit and re-run. 

I guess I can offer the advice to users seeing errors to try exiting and re-running the app repeatedly to see if you get a stable session (I won’t pretend to know why this is behaving inconsistently). 

Hope this is useful and not just making this more maddening to troubleshoot.

**Comment by rickharrison** [comment]
The more descriptive the better. You are seriously amazing. Thanks for posting all this detail.

Can you download 2.106 from the app store and see if that fixes your issues? I just realized it a couple hours ago

**Comment by superfrustrated** [comment]
> The more descriptive the better. You are seriously amazing. Thanks for posting all this detail.
> 
> Can you download 2.106 from the app store and see if that fixes your issues? I just realized it a couple hours ago

So glad I can help. Downloaded latest just now on both devices. I set a reminder for a week (next Saturday) to comment here (if this bug is still open) to say I’ve had no troubles, if that’s the case.  Thanks for jumping on this so quickly!

I really had a laugh at being told I’m seriously amazing for providing so much detail. Not laughing at you, but at the fact that my whole professional life, my managers have consistently been beating me with the Brevity Stick because my communications (email etc) are so detailed; it’s just how I think, and thus how I write. I think I may have discovered that professional bug reporter might have been my calling! 😂

**Comment by superfrustrated** [comment]
@rickharrison in 2.106 build 931, I’ve just experienced the second red error screenshot in my opening report when trying to upvote a comment. I closed the app, reopened it and went back to the same post and the same comment, and I was able to successfully upvote it, and others. I did not have to do any account removal or re-adding or app uninstall. 

I think that whatever the issue or issues are, they’re happening much less frequently than before your fix. 

I realize I’m commenting after running into just one instance of the problem, but like you, I am sure, I was hoping you had managed to fix it, so I thought I’d let you know that there is still something going on making this possible, but that I recovered from it easily.  I guess the point is I just wanted to report that it can still happen, whatever “it“ is.  Thank you for your efforts.

**Comment by superfrustrated** [comment]
@rickharrison I should have thought to try this first and include it above, but trying to switch accounts does not put the app into the sick state right now. It works normally. 

Once upon a time, I was a programmer, and I just have this feeling like there are multiple events or issues that trigger this state, and that you squashed a bunch of them, maybe most of them, but perhaps just not all of them.

**Comment by rickharrison** [comment]
Ok thanks I really appreciate your updates. Please let me know anytime this happens to you

**Comment by rickharrison** [comment]
@superfrustrated Can you join testflight here? https://testflight.apple.com/join/6UJvz6jE I have a new build that I think fixes this issue

**Comment by superfrustrated** [comment]
> @superfrustrated Can you join testflight here? https://testflight.apple.com/join/6UJvz6jE I have a new build that I think fixes this issue

Sure thing, thanks.  Installed on iPhone. Will let you know if I still see the issues. Think I’ll leave the iPad on the current version for contrast, since I use each device a bit most days.

**Comment by rickharrison** [comment]
@superfrustrated 2.107 is also out in the app store now if you want to install that version. Have you run into any issues yet?

**Comment by ilyanep** [comment]
Just wanted to chime in that I did see the upvote bug on 2.106 (which was fixed by switching accounts and switching back) but haven’t yet seen it on 2.107 (sample size like maybe an hour of browsing so far?)

**Comment by superfrustrated** [comment]
> @superfrustrated 2.107 is also out in the app store now if you want to install that version. Have you run into any issues yet?

@rickharrison - Not since I upgraded on the iPhone via TestFlight. You’d already significantly reduced how often this “family” of behaviors was occurring with your previous update, so it’ll take more time to know if you’ve fully squashed them, but it’s looking good! 

I haven’t used the iPad much (running previous version) the last couple of days to offer the contrast I’d hoped to provide you, but I’ll hold off on updating the app on the iPad so that when I do use it, I’ll see if I have the issues there that I hopefully will not be having on the iPhone. 

Feel free to poke me for a report if you care to, maybe after a few more days. Thanks!

**Comment by superfrustrated** [comment]
@rickharrison  Sorry to report, I just encountered the failure to save upvote error on my iPad running v2.108 build 934, iOS 17.1.1. 
Switching to another account and right back to the one I’d just been using let me upvote the same post I’d just been unable to before doing the account switch-and-back. It recovered, in other words, via that little trick. 

No trouble on my iPhone lately, however, I just noticed that it’s running v2.107 build 933, and the version above on the iPad is not being offered for download.  The iPhone is running iOS 16.7.2 (old phone, about to be replaced, not supported by iOS 17, it topped out at iOS 16… the iPad is newer). 

(Sorry, looks like the iPhone was running the older version you offered via TestFlight. Went to App Store and I could download v2.108 build 934, so I’m running the same version on both devices now.) 

Thanks.

**Comment by rickharrison** [comment]
@superfrustrated Build 2.109 is available in testflight. I added a bunch of logging. If you run into errors please upvote a post as that is when the logging will be sent. Can you also send me a screenshot of what you see in the error?

Thanks!

**Comment by superfrustrated** [comment]
@rickharrison got a present for you from v2.109 build 935 on an upvote attempt. Hope you find it useful. It’s a scrollshot, so it’s a tall image but is the full error. 

![IMG_6445](https://github.com/robotswingset/narwhal-bugs/assets/151955564/6aa8351b-ed82-44ae-80ea-6935c1568e42)

**Comment by rickharrison** [comment]
Thanks, I think I finally fixed it for real. New build coming in 15 minutes

**Comment by rickharrison** [comment]
@superfrustrated I deleted that screenshot because it has your reddit token in it. New build in a few

**Comment by superfrustrated** [comment]
> @superfrustrated I deleted that screenshot because it has your reddit token in it. New build in a few

Oi. Thanks for that. This error went with the one you deleted (in case you saved it privately, which would be fine with me). The red error below popped up first, followed by the deleted error on top as the red one slid offscreen. 

![IMG_6446](https://github.com/robotswingset/narwhal-bugs/assets/151955564/b99fac5d-277d-473b-ba82-9d4f065669cd)

**Comment by superfrustrated** [comment]
@rickharrison  Sorry if this is a dumb question, but since I posted the comment with my token in it, which I’m grateful to you for deleting quickly, is that just something that identified my Reddit username? Or would someone fast enough to have snagged it be able to pose as me or something, and would logging out and back in and/or changing mypassword render the token useless, or did I just irreversibly drop my pants?

In other words, I’m unclear how much of a boo boo that was and how to mitigate having posted it, if I can. Thanks.

**Comment by rickharrison** [comment]
To be extra safe, go here and hit revoke access on Narwhal - https://www.reddit.com/prefs/apps/

Then you can re-sign into Narwhal afterwards. Just for the current account you were using

**Comment by rickharrison** [comment]
@superfrustrated Just uploaded 2.109 (937) I think this is the final fix. Please let me know if you run into anything

**Comment by superfrustrated** [comment]
> To be extra safe, go here and hit revoke access on Narwhal - https://www.reddit.com/prefs/apps/
> 
> Then you can re-sign into Narwhal afterwards. Just for the current account you were using

Done. Much obliged.
Going for the new build now.

**Comment by rickharrison** [comment]
@superfrustrated Do you have more than one account signed in to Narwhal?

**Comment by superfrustrated** [comment]
> @superfrustrated Do you have more than one account signed in to Narwhal?

Hi @rickharrison. Yes. Have a total of five logged in on both of my devices. I use my primary 95%+ of the time, an alt maybe 3%, and they others only occasionally to chop away at getting them enough karma so if I decide to switch primary accounts, they'd not at zero. 

Why do you ask? Something you'd like me to test or try? Let me know if so, and what you'd like me to do. 

My app usage had been light the last few days but I have not been seeing issues twisted to this ticket.

**Comment by rickharrison** [comment]
No just wanted to make sure because I think this bug only affected people with 2 or more accounts. Regardless, I think this issue is finally fixed. I'm glad you haven't run into the issues. I think it is finally fixed

**Comment by superfrustrated** [comment]
> No just wanted to make sure because I think this bug only affected people with 2 or more accounts. Regardless, I think this issue is finally fixed. I'm glad you haven't run into the issues. I think it is finally fixed

@rickharrison  That's interesting, but I'm pretty sure that at least though the earlier period where account removal and re-adding was often needed, there were times I only added my primary account back (not all five) and still had the problems you were working on, since I expected to shortly have to repeat  the process.  So perhaps some but not all of the causes were related to having multiple accounts logged in. 

I feel like I'm popping your good idea speculation balloon by telling you that though, since I get how trying to pull patterns or causes out of the noise can help direct efforts or lead you to look at issues in focused ways when you have a theory like that.

App is working well, though again my usage has been light the last few days now. I'll report here if I hit an issue under this bug's umbrella if the bug stays open. 

I think I'll also remove all accounts, the app, reinstall the app, and add only my primary account on my iPad, since I use that for Reddit about half of the time. Maybe it'll lead to something useful with all five on the iPhone and just the one on the iPad (from a clean start). 

Holler if you want an update, but I'll post if I hit an issue.  Thanks much!

**Comment by rickharrison** [comment]
I believe everything with this has been fixed. Thanks for your help debugging :)
